### PR TITLE
Improve performance of  `Dir::inventory()` for many page models

### PR DESCRIPTION
**Pull request by distantnative**
### Enhancements
- Improved  performance of  `Dir::inventory()` for sites with many page models

### Breaking changes
None

## Ready?

- [x] Unit tests for fixed bug/feature
- [x] In-code documentation (wherever needed)
- [x] Tests and checks all pass

### For review team

- [ ] Add changes to release notes draft in Notion
- [x] ~~Add to [website docs release checklist](https://github.com/getkirby/getkirby.com/pulls) (if needed)~~

### CODE REVIEW

**Comment by bastianallgeier** [comment]
This is great. Do you think you can get the code coverage to 100%? Would we put this in 3.9.4?

**Comment by distantnative** [comment]
I'll leave it to the rest of the team whether this should be in 3.9.4 or a later release.

**Review by afbora — COMMENTED** [review]
I tested PR, it seems to work fine 👍 but I can't comment on the code.

**Comment by lukasbestle** [comment]
@distantnative TBH the code changes are quite hard to review because many things changed at once (new internal methods, `switch` refactoring and then the performance improvement). Could you please squash the commits and then split the changes into more meaningful commit chunks to ease the review? Thanks. :)

**Comment by distantnative** [comment]
@lukasbestle I can squash it. But honestly I don't see any sensible good way of splitting that up.

**Comment by distantnative** [comment]
I actually didn't squash it as I see that my commits already represent single steps in the process and not just unclean commits:
1. Refactoring the methods to improve performance at a first step
2. Improve performance of the model lookup for children
3. Unit tests

I tweaked them slightly, but I don't quite understand what more commits you expected @lukasbestle?

File: src/Filesystem/Dir.php [hunk 1]
```diff
@@ -229,131 +229,134 @@ public static function inventory(
...
-		foreach ($content as $contentName) {
-			// could be a meta file. i.e. cover.jpg
-			if (isset($inventory['files'][$contentName]) === true) {
-				continue;
+		// adapt the content extension for mult-language setups
+		if ($multilang === true) {
+			$contentExtension = App::instance()->defaultLanguage()->code() . '.' . $contentExtension;
+		}
+
+		// For many models, one call to scandir() is faster
```
Comment (bastianallgeier) [inline src/Filesystem/Dir.php:312]: Is this still true if the page has a lot of subpages?

**Comment by bastianallgeier** [comment]
I don't think it's harder to read than before. There was always a bit of magic in the inventory methods.

File: src/Filesystem/Dir.php (hunk 1 above)
Comment (distantnative) [reply src/Filesystem/Dir.php:312]: Unclear. I find this really hard to measure and to decide where to focus on performance, it seems to be always a trade off between no of files, templates etc.

**Comment by distantnative** [comment]
@lukasbestle @bastianallgeier do you think we can pursue this in any way? or should we throw this out for now and review when tackling the problem at another time?

**Comment by lukasbestle** [comment]
I cannot review it at the moment. But I don't want to block it at all.

**Comment by distantnative** [comment]
I'm closing it for now so that we have a better overview of PRs we can tackle/handle

**Pull request closed without merging**